### PR TITLE
Read unread branch

### PR DESCRIPTION
**Pull request by dbwest**
I have to get git immersed! That was my first rebase.

don't forget to `rake db:migrate`!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,5 +8,6 @@ script:
   - RAILS_ENV=test bundle exec rake --trace db:migrate
   - bundle exec rake db:test:prepare
   - RAILS_ENV=test bundle exec rake --trace db:seed
+  - bundle exec rake db:migrate
```
Comment (mdarby) [inline .travis.yml:11]: We're already doing this in line #8

File: Gemfile [hunk 2]
```diff
@@ -20,6 +20,8 @@ gem 'omniauth'
 gem 'omniauth-twitter'
 gem 'omniauth-facebook'
 
+gem 'unread'
```
Comment (mdarby) [inline Gemfile:23]: Just being nitpicky -- but could you please keep the gems listed together.

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@ class CommentsController < ApplicationController
   respond_to :js
 
   def new
-    @conversation = Conversation.find(params[:conversation_id])
+    @conversation = Conversation.forind(params[:conversation_id])
```
Comment (mdarby) [inline app/controllers/comments_controller.rb:5]: Typo?

File: app/controllers/conversations_controller.rb [hunk 4]
```diff
@@ -1,8 +1,11 @@
 class ConversationsController < ApplicationController
-  before_filter :authenticate_user!, except: [:show]
+  before_filter :authenticate_user!
```
Comment (mdarby) [inline app/controllers/conversations_controller.rb:2]: Why was this removed? We want public access to a Conversation.

File: app/controllers/conversations_controller.rb [hunk 5]
```diff
@@ -1,8 +1,11 @@
...
 
   def show
     @conversation  = Conversation.find(params[:id], include: {:comments => [:user, :uploaded_files]})
+    @conversation.comments.each do |c| 
+      c.mark_as_read! :for => current_user
+    end
```
Comment (mdarby) [inline app/controllers/conversations_controller.rb:8]: Ah, I see, you removed the `except: [:show]` part as you needed a `current_user` in `Conversation#show`. All good, but we can't rely on `current_user` being available as this might be publicly viewed.

File: app/controllers/forums_controller.rb [hunk 6]
```diff
@@ -1,4 +1,5 @@
 class ForumsController < ApplicationController
+  before_filter :authenticate_user!
```
Comment (mdarby) [inline app/controllers/forums_controller.rb:2]: Same with this -- it should be public

File: app/models/admin_user.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  # Include default devise modules. Others available are:
+  # :token_authenticatable, :confirmable,
+  # :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, 
+         :recoverable, :rememberable, :trackable, :validatable
+
+  # Setup accessible (or protected) attributes for your model
+  attr_accessible :email, :password, :password_confirmation, :remember_me
+  # attr_accessible :title, :body
+end
```
Comment (mdarby) [inline app/models/admin_user.rb:11]: I do believe we already have an `AdminUser` that is given to us by ActiveAdmin?

File: app/views/conversations/_conversation.html.slim [hunk 8]
```diff
@@ -2,3 +2,5 @@ tr[id="conversation_#{conversation.id}"]
   td= link_to(conversation.title, conversation_path(conversation))
   td= pluralize(conversation.comments_count, "Comment")
   td= timeago conversation.try(:updated_at).try(:utc)
+  td
+    .span1.label.label-info= "#{conversation.comments.unread_by(current_user).count} unread"
```
Comment (mdarby) [inline app/views/conversations/_conversation.html.slim:6]: Please match the style of the other `td` elements. No need for the classes.

File: db/migrate/20121002174002_devise_create_admin_users.rb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+      t.timestamps
+    end
+
+    add_index :admin_users, :email,                :unique => true
+    add_index :admin_users, :reset_password_token, :unique => true
+    # add_index :admin_users, :confirmation_token,   :unique => true
+    # add_index :admin_users, :unlock_token,         :unique => true
+    # add_index :admin_users, :authentication_token, :unique => true
+  end
+end
```
Comment (mdarby) [inline db/migrate/20121002174002_devise_create_admin_users.rb:52]: This is not necessary as ActiveAdmin contains an AdminUser.

**Comment by mdarby** [comment]
I've made some (hopefully constructive) notes :)

**Comment by skord** [comment]
We don't use admin user. It's one of the first things I did. We use the regular old devise and an admin role.

File: db/migrate/20121002174002_devise_create_admin_users.rb (hunk 9 above)
Comment (dbwest) [inline db/migrate/20121002174002_devise_create_admin_users.rb:52]: thanks, I couldn't find it. I was using that, and then:

```
rails c
User.create(id: 1, email: "foo@bar.org", password: "password")
user_foo = User.find(1)
user_foo.update_attribute :role, "admin"
```

to make an admin I could use for testing

What's the default admin and pass?

File: app/views/conversations/_conversation.html.slim (hunk 8 above)
Comment (dbwest) [inline app/views/conversations/_conversation.html.slim:6]: That makes sense.

File: db/migrate/20121002174002_devise_create_admin_users.rb (hunk 9 above)
Comment (mdarby) [inline db/migrate/20121002174002_devise_create_admin_users.rb:52]: admin@example.com / password

File: app/models/admin_user.rb (hunk 7 above)
Comment (dbwest) [inline app/models/admin_user.rb:11]: I tried what I thought was the default admin:
admin@example.com
password

didn't work for me
I was following the instructions from
http://activeadmin.info/documentation.html

Comment (mdarby) [inline app/models/admin_user.rb:11]: Did you run `rake db:seed`? That installs a bunch of default data.

File: app/controllers/conversations_controller.rb (hunk 5 above)
Comment (dbwest) [inline app/controllers/conversations_controller.rb:8]: I guess that was a silly kludge. It can work without that change, but I wasn't sure how to do it better.

File: app/controllers/comments_controller.rb (hunk 3 above)
Comment (dbwest) [inline app/controllers/comments_controller.rb:5]: yes.

File: Gemfile (hunk 2 above)
Comment (dbwest) [inline Gemfile:23]: sure. not too nitpicky

File: .travis.yml (hunk 1 above)
Comment (dbwest) [inline .travis.yml:11]: I don't understand the Travis stuff at all, honestly. I tried running my own fork off of Travis, and it wasn't building so I messed with that. I could have removed my edit of this file from git staging.

Comment (mdarby) [inline .travis.yml:11]: Travis is what we call a "Continuous Integration" server, and what that does is simply install, build, and run our tests every time someone pushes to master (in the main repo), or submits a pull request. You don't need to run it locally at all -- just run `rake` to run all the tests locally. :)

File: app/controllers/conversations_controller.rb (hunk 5 above)
Comment (mdarby) [inline app/controllers/conversations_controller.rb:8]: The code is fine, but just do a check first -- `if current_user.present?`, then run your block.

File: app/models/admin_user.rb (hunk 7 above)
Comment (dbwest) [inline app/models/admin_user.rb:11]: thanks!

File: app/controllers/conversations_controller.rb (hunk 4 above)
Comment (dbwest) [inline app/controllers/conversations_controller.rb:2]: part of the kludge mentioned below

File: .travis.yml (hunk 1 above)
Comment (dbwest) [inline .travis.yml:11]: ahh. first non-book encounter with CI
